optuna/samplers/_base.py

```diff
@@ -137,6 +140,37 @@ def sample_independent(
 
         raise NotImplementedError
 
+    def after_trial(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        state: TrialState,
+        values: Optional[Sequence[float]],
+    ) -> None:
+        """Trial post-processing.
+
+        This method is called after the objective function returns and right before the trials is
+        finished and its state is stored.
+
+        .. note::
+            Added in v2.4.0 as an experimental feature. The interface may change in newer versions
+            without prior notice. See https://github.com/optuna/optuna/releases/tag/v2.4.0.
```

Comment:
Avoided using the `@experimental` decorator since this method is called internally and we don't want actual warnings being propagated to the user. We can silence warning internally upon invocations but given that particular samplers could raise actual warnings, might not be very smart.